src/index.ts

```diff
@@ -1,4 +1,20 @@
-import curryToArity from "./util.js";
+import curryToArity from "./util";
+
+type Rule = {
```

Comment:
Nice to have this is TS. :+1:
Maybe as a futre step type could be refined a bit using generics with something like this:

```ts
type Rule<Facts, Result, Input> =
  Plainrule<Facts, Result, Input> |
  InjectedRule<Facts, Result, Input> |
  TransformedRule<Facts, Result, Input> |
  FirstRule<Facts, Result, Input> |
  AllRule<Facts, Result, Input> |
  ChainRule<Facts, Result, Input> |
  IfRule<Facts, Result, Input>;

type PlainRule<Facts, Result, Input> = {
  matcher: (facts: Facts, Input: input) => boolean,
  action: (facts: Facts, Input: input) => Result,
}

// ... add types for other rule types here

type RuleResult<Value> = {
  foundMatch: boolean;
  value: Value;
};
```

Reply:
Fixed it :)

Reply:
Awesome looks good :+1:

One more small suggestion: Maybe if possible, instead of defaulting to `any` in the type signatures like `Facts = any` we could use `Facts = unknown`. That way we force the user to pass a type and errors will be caught quicker at build time.